### DISCUSSION TITLE
Define maximum query depth for graphQL?

### DISCUSSION THREAD

**Discussion by fredro84** [supabase/supabase#29796] [Questions]
Recently I have been migrating all of my graphql queries from a custom postgraphile api to the supabase graphql api. During the migration I was investigating which safety measures I can implement against large (accidental) queries. I found that the max number of rows returned is 30 and totalCount is disabled by default. But I can't find whether I can configure the maximum query depth for a graphql query. This means 30 rows for each level of the query, but if that query can go 20 levels deep, the response will be huge.

I found the default timeout settings for users: 3 sec (anon), 8 sec (authenticated). But for an authenticated user they can probably use up quite some resources in 8 seconds if I don't limit the query depth. 

Does anyone know what a good approach is here?

**Comment by sand97** [comment]
Hi @fredro84 any update ?

**Reply by fredro84** [reply]
@sand97 defining query depth is still not possible and I find the solution a little too risky what @steve-chavez proposes below. So I'm slowly migrating away from graphql all together. Supabase doesn't supply the required security measures at the moment. Because rate limiting still hasn't been developed, this means that any (authenticated) user could still send multiple heavy queries that would put heavy strain on my database, even if I have set up the cost_limit for one request. 

What I have decided to do instead is create custom SQL functions that return a fixed set of data, and I limit the amount of results that are returned specifically per function. I call these functions with the supabase rpc method. It is less flexible than graphql, but I see no other option to build something secure enough for now. A bit disappointed though, because I was a big fan of graphQL, especially with a PostGraphile setup. 

But don't get me wrong, I still think Supabase is by far the best value for money!

**Comment by steve-chavez** [comment]
You can use the [pg_plan_filter](https://github.com/pgexperts/pg_plan_filter) module to limit the query cost, this is included in Supabase. For example, you could do:

```
ALTER ROLE authenticated SET plan_filter.statement_cost_limit = 100000.0;
```

**Comment by damassi** [comment]
I just ran into this issue as well, using the graphql API. I'm defining a chat interface and right now the maximum number of rows returned is 30 for my messages, even though I have `first: 100` set in my graphql connection query. 

EDIT: I just answered my own question! This needs to be set at the pg_graphql level: https://supabase.github.io/pg_graphql/configuration/#max-rows
